### PR TITLE
Add tests for FIPS keygen self test failures.

### PR DESCRIPTION
**Pull request by slontis**
During key generation RSA, EC and DSA have extra tests that run in FIPS mode All 3 algorithms have a pairwise test, EC & DSA also run a KAT test.

This test uses the self test callback to force an error during each of the extra pairwise and KAT tests.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — DISMISSED** [review]
You are so negative these days...

**Comment by paulidale** [comment]
I'd consider this for 3.1 as well.

**Comment by slontis** [comment]
Have to make changes as the KATs tests were added for 3.1

File: test/pairwise_fail_test.c [hunk 1]
```diff
@@ -19,6 +19,7 @@ typedef enum OPTION_choice {
     OPT_CONFIG_FILE,
     OPT_PAIRWISETEST,
     OPT_DSAPARAM,
+    OPT_FIPSVERSION,
```
Comment (paulidale) [inline test/pairwise_fail_test.c:22]: Elsewhere we do the version test in the .t file and skip not the C.  Any reason for doing it this way instead?

Comment (slontis) [reply test/pairwise_fail_test.c:22]: Can you point me to an example in a .t?

Comment (paulidale) [reply test/pairwise_fail_test.c:22]: It's something like this in the SKIP section:
```perl
    run(test(["fips_version_test", "-config", $provconf, ">=3.1.0"]),
             capture => 1, statusvar => \my $exit);
    skip "FIPS provider version too old", 1
        if !$exit;
```
A bit of setup is required: https://github.com/openssl/openssl/blob/e788c772b12eea5ced4ce46619e13acf0e0eb6ba/test/recipes/25-test_verify.t#L355-L367

Fortunately, the setup and check only needs to be done once and can be used in both places you require it.

File: test/recipes/30-test_pairwise_fail.t [hunk 2]
```diff
@@ -23,34 +23,43 @@ plan skip_all => "These tests are unsupported in a non fips build"
     if disabled("fips");
 
 plan tests => 5;
+my $provconf = srctop_file("test", "fips-and-base.cnf");
+
+run(test(["fips_version_test", "-config", $provconf, ">=3.1.0"]),
+    capture => 1, statusvar => \my $exit);
```
Comment (paulidale) [inline test/recipes/30-test_pairwise_fail.t:29]: `$fips_exit` might be more understandable.

**Review by paulidale — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
> Have to make changes as the KATs tests were added for 3.1

I do not understand, won't this apply to 3.1 cleanly?

**Review by t8m — APPROVED** [review]
OK also for 3.1 if it applies there.

**Comment by paulidale** [comment]
IMO, we definitely want this in 3.1 too.  If it needs to be a separate PR, so be it.

**Comment by slontis** [comment]
> > Have to make changes as the KATs tests were added for 3.1
> 
> I do not understand, won't this apply to 3.1 cleanly?

I just meant FIPSVersion checks needed to be added in order for the CI to pass the backwards compatible tests. It should be ok for 3.1.

**Comment by paulidale** [comment]
It's probably good for 3.0 too.  Even though it's a test, I think keeping these aligned is good because we test against various FIPS provider versions (both forwards and backward).

**Comment by t8m** [comment]
Merged to master branch. Thank you for your contribution.

This does not cherry-pick cleanly to 3.1, if you want this on 3.1, and 3.0, please submit a backport PR.

**Pull request closed without merging**